ironfish/src/blockchain/nullifierSet/nullifierSet.ts

```diff
@@ -0,0 +1,132 @@
+  value: NullifierInfo
+}
+
+export class NullifierSet {
+  readonly db: IDatabase
+
+  // Keep track of number of nullifiers in the set
+  private readonly counter: IDatabaseStore<CounterEntry<'Size'>>
+
+  // Nullifier -> BlockHash, TransactionHash
+  private readonly nullifiers: IDatabaseStore<NullifiersSchema>
+
+  constructor(options: { db: IDatabase; storeName: string }) {
+    this.db = options.db
+
+    this.counter = this.db.addStore({
+      name: `${options.storeName}c`,
+      keyEncoding: new StringEncoding<'Size'>(),
+      valueEncoding: U32_ENCODING,
+    })
+
+    // Nullifier -> BlockHash, TransactionHash
+    this.nullifiers = this.db.addStore({
+      name: options.storeName,
+      keyEncoding: BUFFER_ENCODING,
+      valueEncoding: new NullifierLocationEncoding(),
+    })
+  }
+
+  size(tx?: IDatabaseTransaction): Promise<number> {
+    return this.db.withTransaction(tx, async (tx) => {
+      const size = await this.counter.get('Size', tx)
+      return size === undefined ? 0 : size
+    })
+  }
+
+  contains(nullifier: Nullifier, tx?: IDatabaseTransaction): Promise<boolean> {
+    return this.db.withTransaction(tx, async (tx) => {
+      return await this.nullifiers.has(nullifier, tx)
+    })
+  }
+
+  contained(nullifier: Nullifier, size: number, tx?: IDatabaseTransaction): Promise<boolean> {
+    return this.db.withTransaction(tx, async (tx) => {
+      const nullifierInfo = await this.nullifiers.get(nullifier, tx)
+      if (nullifierInfo === undefined) {
+        return false
+      }
+
+      return nullifierInfo.position < size
+    })
+  }
+
+  get(nullifier: Nullifier, tx?: IDatabaseTransaction): Promise<NullifierInfo | undefined> {
+    return this.db.withTransaction(tx, async (tx) => {
+      return await this.nullifiers.get(nullifier, tx)
+    })
+  }
+
+  connectBlock(transactions: Transaction[], tx?: IDatabaseTransaction): Promise<void> {
+    return this.db.withTransaction(tx, async (tx) => {
+      let currentSize = await this.size(tx)
+
+      for (const transaction of transactions) {
+        for (const spend of transaction.spends()) {
+          // Throws an error if a nullifier already exists
+          // We should never allow overwriting a nullifier
+          await this.nullifiers.add(
+            spend.nullifier,
+            {
+              transactionHash: transaction.hash(),
+              position: currentSize,
+            },
+            tx,
+          )
+
+          currentSize++
+        }
+      }
+
+      await this.counter.put('Size', currentSize, tx)
+    })
+  }
+
+  disconnectBlock(transactions: Transaction[], tx?: IDatabaseTransaction): Promise<void> {
+    return this.db.withTransaction(tx, async (tx) => {
+      let currentSize = await this.size(tx)
+
+      for (const transaction of transactions) {
+        for (const spend of [...transaction.spends()].reverse()) {
+          // Sanity check that we are removing the correct block
+          // Not necesarrally needed but
```

Comment:
I think this comment needs finishing
```suggestion
          // Not necessarily needed but
```

Reply:
thanks, fixed